### PR TITLE
nixos/nginx: add proxyCache options

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
 - Changed default paths for temporary files to `/tmp/nginx_*`.
 - Added proxyCache options.

cc @dasJ @ajs124 @SuperSandro2000

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 1]
```diff
@@ -689,6 +693,71 @@ in
...
+              description = lib.mdDoc "Set size to shared memory zone.";
+            };
+
+            levels = mkOption {
+              type = types.str;
+              default = "1:2";
+              example = "1:2:2";
+              description = lib.mdDoc ''
+                The levels parameter defines hierarchy levels of a cache: from 1 to 3,
+                each level accepts values 1 or 2.
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:721]: This doesn't explain what the difference between 2 and 3 numbers. Can we simply link the upstream doc?

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 2]
```diff
@@ -689,6 +693,71 @@ in
...
+            };
+
+            useTempPath = mkOption {
+              type = types.bool;
+              default = false;
+              example = true;
+              description = lib.mdDoc ''
+                Nginx first writes files that are destined for the cache to a temporary
+                storage area, and the use_temp_path=off directive instructs Nginx to
+                write them to the same directories where they will be cached. Recommend
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:732]:
```suggestion
                write them to the same directories where they will be cached. Recommended
```

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 3]
```diff
@@ -689,6 +693,71 @@ in
...
+              description = lib.mdDoc ''
+                Cached data that has not been accessed for the time specified by
+                the inactive parameter is removed from the cache, regardless of
+                its freshness.
+              '';
+            };
+
+            maxSize = mkOption {
+              type = types.nullOr types.str;
+              default = null;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:751]: I think we want some sane default here like 1GB

File: pkgs/servers/http/nginx/generic.nix [hunk 4]
```diff
@@ -79,11 +79,11 @@ stdenv.mkDerivation {
...
-    "--http-client-body-temp-path=/var/cache/nginx/client_body"
-    "--http-proxy-temp-path=/var/cache/nginx/proxy"
-    "--http-fastcgi-temp-path=/var/cache/nginx/fastcgi"
-    "--http-uwsgi-temp-path=/var/cache/nginx/uwsgi"
-    "--http-scgi-temp-path=/var/cache/nginx/scgi"
+    "--http-client-body-temp-path=/tmp/nginx_client_body"
+    "--http-proxy-temp-path=/tmp/nginx_proxy"
+    "--http-fastcgi-temp-path=/tmp/nginx_fastcgi"
+    "--http-uwsgi-temp-path=/tmp/nginx_uwsgi"
+    "--http-scgi-temp-path=/tmp/nginx_scgi"
```
Comment (SuperSandro2000) [inline pkgs/servers/http/nginx/generic.nix:86]: Is this path sandboxed in our systemd unit?

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 5]
```diff
@@ -192,6 +192,10 @@ let
 
       server_tokens ${if cfg.serverTokens then "on" else "off"};
 
+      ${optionalString (cfg.proxyCache.enable) ''
+        proxy_cache_path /var/cache/nginx keys_zone=${cfg.proxyCache.keysZoneName}:${cfg.proxyCache.keysZoneSize} levels=${cfg.proxyCache.levels} use_temp_path=${if cfg.proxyCache.useTempPath then "on" else "off"} ${optionalString (cfg.proxyCache.inactive != null) "inactive=${cfg.proxyCache.inactive}"} ${optionalString (cfg.proxyCache.maxSize != null) "max_size=${cfg.proxyCache.maxSize}"};
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:196]: Please split this line into multiple lines.

Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:196]: If split lines, then in end the nginx configuration will display like this:
```
  proxy_cache_path /var/cache/nginx/proxy_cache levels=1:2
  keys_zone=CACHE:32m max_size=1g inactive=60m;
  proxy_ignore_headers "Set-Cookie";
  server {
    listen 0.0.0.0:80 ;
    listen [::0]:80 ;
```
Difficult to view configuration.
Can't generate this variant:
```
  proxy_cache_path /var/cache/nginx/proxy_cache levels=1:2
                                keys_zone=CACHE:32m max_size=1g inactive=60m;
  proxy_ignore_headers "Set-Cookie";
  server {
    listen 0.0.0.0:80 ;
    listen [::0]:80 ;
```

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 1 above)
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:721]: Description taken from here - http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path
```
Sets the path and other parameters of a cache. Cache data are stored in files. The file name in a cache is a result of applying the MD5 function to the [cache key](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_key). The levels parameter defines hierarchy levels of a cache: from 1 to 3, each level accepts values 1 or 2. For example, in the following configuration
```

File: pkgs/servers/http/nginx/generic.nix (hunk 4 above)
Comment (Izorkin) [reply pkgs/servers/http/nginx/generic.nix:86]: Yes.

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 1 above)
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix]: Update

**Comment by RaitoBezarius** [comment]
Can we configure only one `proxy_cache_path` ?

**Comment by Izorkin** [comment]
> Can we configure only one `proxy_cache_path` ?

Are you talking about this variant?:
```
  proxy_cache_path /var/cache/nginx keys_zone=${cfg.extraOptions}
```

Advanced settings are needed so that can flexibly configure caching for a virtual host:
```
        locations."@proxy" = {
...
          extraConfig = ''
            ${nginxProxyHeaders}
            proxy_pass_header Server;
            proxy_buffering on;
            proxy_redirect off;
            proxy_cache ${config.services.nginx.proxyCache.keysZoneName};
            proxy_cache_valid 200 7d;
            proxy_cache_valid 410 24h;
...
```

**Comment by RaitoBezarius** [comment]
> > Can we configure only one `proxy_cache_path` ?
> 
> Are you talking about this variant?:
> 
> ```
>   proxy_cache_path /var/cache/nginx keys_zone=${cfg.extraOptions}
> ```
> 
> Advanced settings are needed so that can flexibly configure caching for a virtual host:
> 
> ```
>         locations."@proxy" = {
> ...
>           extraConfig = ''
>             ${nginxProxyHeaders}
>             proxy_pass_header Server;
>             proxy_buffering on;
>             proxy_redirect off;
>             proxy_cache ${config.services.nginx.proxyCache.keysZoneName};
>             proxy_cache_valid 200 7d;
>             proxy_cache_valid 410 24h;
> ...
> ```

Yes, I mean, it's fine, I just want to avoid some setup that would prevent configuring multiple zone names for different caches in the same NGINX, but I do not know if that's possible at all.

**Comment by vcunat** [comment]
My bisection says that the path-changing commit broke a test and is blocking `nixos-unstable` channel: https://hydra.nixos.org/build/200295215

**Comment by vcunat** [comment]
Test maintainers: @adisbladis, @saschagrunert, @vdemeester, @zowoq.

I retried a couple times on two machines, and it seems well reproducible.  (Beware that the failure is a 15-minute timeout, but success would be immediate.)

**Comment by SuperSandro2000** [comment]
```
podman # [  968.716418] nginx[80862]: nginx: [emerg] mkdir() "/tmp/nginx_client_body" failed (2: No such file or directory)
```

**Comment by balsoft** [comment]
Should I revert?

**Comment by de2tla2f** [comment]
oh-no 😰🕯️

**Comment by Izorkin** [comment]
Working this patch?:
```diff
diff --git a/pkgs/build-support/docker/examples.nix b/pkgs/build-support/docker/examples.nix
index 1e9f07045e3..ed76e46d97e 100644
--- a/pkgs/build-support/docker/examples.nix
+++ b/pkgs/build-support/docker/examples.nix
@@ -94,6 +94,7 @@ rec {
     extraCommands = ''
       # nginx still tries to read this directory even if error_log
       # directive is specifying another file :/
+      mkdir -p tmp
       mkdir -p var/log/nginx
       mkdir -p var/cache/nginx
     '';

```

**Comment by RaitoBezarius** [comment]
Why are we using explicit /tmp rather than using systemd features?

**Comment by RaitoBezarius** [comment]
> > Why are we using explicit /tmp rather than using systemd features?
> 
> By default `PrivateTmp` option in systemd is used, which creates an isolated /tmp folder. Container does not use systemd, so have to create directory `tmp` manually.

Hmm, NixOS containers do not have a systemd PID1 or something? How can the service even start? It depends so much on systemd.

I do think that `/tmp` is not the question but why are we not using `RuntimeDirectory` and putting the cache into `/run/nginx/`?

**Comment by Izorkin** [comment]
> Hmm, NixOS containers do not have a systemd PID1 or something? How can the service even start? It depends so much on systemd.
> 
> I do think that `/tmp` is not the question but why are we not using `RuntimeDirectory` and putting the cache into `/run/nginx/`?

Systemd service is not used to run nginx. Moving to `/run/nginx/` folder will not fix error.

**Comment by RaitoBezarius** [comment]
> > Hmm, NixOS containers do not have a systemd PID1 or something? How can the service even start? It depends so much on systemd.
> > I do think that `/tmp` is not the question but why are we not using `RuntimeDirectory` and putting the cache into `/run/nginx/`?
> 
> Systemd service is not used to run nginx. Moving to `/run/nginx/` folder will not fix error.

I see it now, anyway, we should not be using `/tmp` on systemd-enabled and reading the NGINX container service, it seems like it follow systemd directory structure by creating `var/log/nginx`, so I guess a technical debt cleanup would involve to do `run/nginx` too.

At some point, we should be able to derive these folders to create given the systemd unit for example too, removing the need for doing it manually and thinking of possible paths.

**Comment by Izorkin** [comment]
> I see it now, anyway, we should not be using `/tmp` on systemd-enabled and reading the NGINX container service, it seems like it follow systemd directory structure by creating `var/log/nginx`

Needed completely remove nginx directories from `/tmp`?

> so I guess a technical debt cleanup would involve to do `run/nginx` too.

Needed create a directory `run/nginx` in `pkgs/build-support/docker/examples.nix`?

**Comment by RaitoBezarius** [comment]
> > I see it now, anyway, we should not be using `/tmp` on systemd-enabled and reading the NGINX container service, it seems like it follow systemd directory structure by creating `var/log/nginx`
> 
> Needed completely remove nginx directories from `/tmp`?

Yes.

> 
> > so I guess a technical debt cleanup would involve to do `run/nginx` too.
> 
> Needed create a directory `run/nginx` in `pkgs/build-support/docker/examples.nix`?

Yes.

**Comment by Izorkin** [comment]
> > Needed completely remove nginx directories from `/tmp`?
> 
> Yes.

Any possible variants?
These directories are already in use:
 - `/run/nginx` - located is PID file;
 - `/var/cache/nginx` - located cached files;
 - `/var/log/nginx` - located log files;
Directory locations in `proxy_temp_path` and `proxy_cache_path` must be different.

> > Needed create a directory `run/nginx` in `pkgs/build-support/docker/examples.nix`?
> 
> Yes.

In `pkgs/build-support/docker/examples.nix` there is no need to create directory `/run/nginx`, because location of PID file in configuration is set in `/dev/null` - `pid /dev/null;`

File: pkgs/servers/http/nginx/generic.nix (hunk 4 above)
Comment (misuzu) [reply pkgs/servers/http/nginx/generic.nix:86]: This change breaks my workflow™
I'm serving temporary files from `/tmp/data` and have `systemd.services.nginx.serviceConfig.PrivateTmp = mkForce false;` to make this work. After this change nginx fails to start:
```
Dec 06 12:37:03 nixos nginx-pre-start[2881]: nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
Dec 06 12:37:03 nixos nginx-pre-start[2881]: nginx: [emerg] mkdir() "/tmp/nginx_client_body" failed (30: Read-only file system)
Dec 06 12:37:03 nixos nginx-pre-start[2881]: nginx: configuration file /etc/nginx/nginx.conf test failed
Dec 06 12:37:03 nixos systemd[1]: nginx.service: Control process exited, code=exited, status=1/FAILURE
```

Comment (Izorkin) [reply pkgs/servers/http/nginx/generic.nix:86]: @misuzu try add to configuration:
```
  systemd.services.nginx.serviceConfig.ReadWritePaths = [ "/tmp" ];
```

Comment (misuzu) [reply pkgs/servers/http/nginx/generic.nix:86]: @Izorkin this fixes my issue, thank you!

**Pull request merged**